### ISSUE TITLE
Detectron and Caffe2 Inference result are not same for same model and  input image .

### ISSUE THREAD

**Issue by ParoRahul** [facebookresearch/Detectron#921]
I trained A Detectron Model with my custom dataset.
Model Name - e2e_faster_rcnn_R-101-FPN.yaml
number of class in Dataset - 59

it worked completely fine.

then i convert the model in caffe2 with script "convert_pkl_to_pb.py -- device 'cpu'  "
it generates 3 file 
1.model_init.pb
2.model.pb
3.param_init_net.pbtxt

then I go for inference with caffe2. script for inference is given bellow

### start of the script 

from caffe2.proto import caffe2_pb2
from caffe2.python import core, workspace
from caffe2.python import dyndep
from caffe2.python import scope

import os
import numpy as np 
import cv2
import csv

def sort_results(boxes, segms, keypoints, classes):
    indices = np.argsort(boxes[:, -1])[::-1]
    if boxes is not None:
        boxes = boxes[indices, :]
    if segms is not None:
        segms = [segms[x] for x in indices]
    if keypoints is not None:
        keypoints = [keypoints[x] for x in indices]
    if classes is not None:
        if isinstance(classes, list):
            classes = [classes[x] for x in indices]
        else:
            classes = classes[indices]
    return boxes, segms, keypoints, classes

def get_device_option_cpu():
    device_option = core.DeviceOption(caffe2_pb2.CPU)
    return device_option    

def prepare_blobs(image_Path,pixel_means,target_size,max_size):
    '''    Reference: blob.prep_im_for_blob()   '''
    im = cv2.imread(image_Path)   
    im = im.astype(np.float32, copy=False)
    im -= pixel_means
    im_shape = im.shape
    im_size_min = np.min(im_shape[0:2])
    im_size_max = np.max(im_shape[0:2])
    im_scale = float(target_size) / float(im_size_min)
    if np.round(im_scale * im_size_max) > max_size:
       im_scale = float(max_size) / float(im_size_max)
    im = cv2.resize(im,None,None,fx=im_scale,fy=im_scale,interpolation=cv2.INTER_LINEAR)
    blob = im_list_to_blob(im)
    blobs = {}
    blobs['data'] = blob
    blobs['im_info'] = np.array(
        [[blob.shape[2], blob.shape[3], im_scale]],
        dtype=np.float32
    )
    return blobs,im_shape[0],im_shape[1]

def im_list_to_blob(ims):
    """Convert a list of images into a network input. Assumes images were
    prepared using prep_im_for_blob or equivalent: i.e.
      - BGR channel order
      - pixel means subtracted
      - resized to the desired input size
      - float32 numpy ndarray format
    Output is a 4D HCHW tensor of the images concatenated along axis 0 with
    shape.
    """
    if not isinstance(ims, list):
        ims = [ims]
    max_shape = np.array([im.shape for im in ims]).max(axis=0)
    # Pad the image so they can be divisible by a stride
    if True:
        stride = float(32)
        max_shape[0] = int(np.ceil(max_shape[0] / stride) * stride)
        max_shape[1] = int(np.ceil(max_shape[1] / stride) * stride)
    num_images = len(ims)
    blob = np.zeros((num_images, max_shape[0], max_shape[1], 3), dtype=np.float32)
    for i in range(num_images):
        im = ims[i]
        blob[i, 0:im.shape[0], 0:im.shape[1], :] = im
    # Move channels (axis 3) to axis 1
    # Axis order will become: (batch elem, channel, height, width)
    channel_swap = (0, 3, 1, 2)
    blob = blob.transpose(channel_swap)
    return blob    

ModelPath='D:/Model_data/Path/cpu'
IMAGE_LOCATION='D:/Img/Loaction/'
OUTPUTPATH='D:/outputpath/'
PIXEL_MEANS = np.array([[[102.9801, 115.9465, 122.7717]]])
TEST_SCALE = 600
TEST_MAX_SIZE = 1000

dyndep.InitOpsLibrary('C:/ProgramData/Anaconda3/envs/Pytorch/Lib/site-packages/torch/lib/caffe2_detectron_ops.dll')
init_net = caffe2_pb2.NetDef()
predict_net = caffe2_pb2.NetDef()
with open(os.path.join(ModelPath,"model_init.pb"),'rb') as f:
     init_net.ParseFromString(f.read())
with open(os.path.join(ModelPath,"model.pb"),'rb') as f:
     predict_net.ParseFromString(f.read())
workspace.ResetWorkspace()
workspace.RunNetOnce(init_net)
print(type(predict_net),predict_net.name)
for op in predict_net.op:
    for blob_in in op.input:
        if not workspace.HasBlob(blob_in):
            workspace.CreateBlob(blob_in)
print('Opreators Are Loaded')
workspace.CreateNet(predict_net)
print('Predictor Net Created')
imList=[]
if os.path.isdir(IMAGE_LOCATION):
    image_list=[ os.path.join(IMAGE_LOCATION,file) for file in os.listdir(IMAGE_LOCATION)
                if file.split('.')[1] == 'jpg' ]
else:
    image_list=[IMAGE_LOCATION]
outputList,printList=[],[]    
for test_imag in image_list:
    input_blobs,height,width = prepare_blobs(test_imag,PIXEL_MEANS,TEST_SCALE,TEST_MAX_SIZE)
    print('inputs Bloobs Are Preapared')
    gpu_blobs = []
    for k, v in input_blobs.items():
        workspace.FeedBlob(core.ScopedName(k),v,
        get_device_option_cpu())
    try:
        workspace.RunNet(predict_net.name)
        scores = workspace.FetchBlob('score_nms')
        classids = workspace.FetchBlob('class_nms')
        boxes = workspace.FetchBlob('bbox_nms')
        #print(type(boxes),type(scores),type(classids))
        #print(boxes.shape,scores.shape,classids.shape,input_blobs['im_info'])
        outputList.append({'file':os.path.basename(test_imag),
                           'boxex':boxes,
                           'score':scores,
                           'classId':classids})
        for (box, score, classID) in zip(boxes,scores,classids):
            if score < 0.9 :
               continue 
            xmin=int(float(box[0]))
            ymin=int(float(box[1]))
            xmax=int(float(box[2]))
            ymax=int(float(box[3]))
            classID=int(classID)
            score=round(score,4)
            printList.append([os.path.basename(test_imag),height,width,classID,xmin,ymin,xmax,ymax,score])                   
    except Exception as e :
        print('Running pb model failed.\n{}'.format(e))
# Writing Result in CSV Format        
fileNameCsv='Caff2_cpu_Windows_Data.csv'
AnnoCSVPath=os.path.join(OUTPUTPATH,fileNameCsv)
try:
    with open(AnnoCSVPath,'w') as annotation_file :
         writer = csv.writer(annotation_file,delimiter=',')
         writer.writerow(['filename','width','height','class','xmin','ymin','xmax','ymax','score'])
         writer.writerows(printList)
except Exception as e:
       print (e)       

### End of scripts of script

There id big mismatch in the result of Caffe2 and Detectron inference. 

Number of bounding box predicted in both the Image was same 
bounding box label is even the same
But Score of the bounding box is not the same, not even approximately equal

###Caffe2 Result 

![20180412_104410](https://user-images.githubusercontent.com/32712591/61696180-831ab580-ad52-11e9-948a-24cb78b897ca.jpg)

* Operating system: Ubuntu
* Compiler version: gcc version 7.4.0 
* CUDA version:  10.1
* cuDNN version: 7.5.0
* NVIDIA driver version: 418.40.04 
* GPU models (for all devices if they are not all the same): Tesla M60
* `python --version` output: Python 3.6.7
